Improve mirror selection logic in Closest with full-server fallback

SuperKali opened the pull request:
This PR enhances the `Closest` function used for mirror selection by ensuring that **all** configured servers are considered, even when the initial filtering (based on availability, protocol, and rule checks) is too restrictive. It introduces a fallback mechanism, and more robust local mirror selection logic.

**Fixes:** Problems detecting nearest servers #15 

## Changes

- **Fallback Mechanism:**
  - If the filtered list of valid servers (i.e., servers that are available, support the required protocol, and pass rule checks) contains fewer than two mirrors, the function now falls back to using the full server list.
  - This ensures that all configured mirrors are considered during mirror selection.

- **Local Mirror Selection:**
  - After obtaining valid servers, the function filters for local servers (mirrors with the same country code as the client).
  - If at least one local mirror is found:
    - The distances are computed for each local mirror.
    - If the nearest local mirror is within a defined threshold (200 km), it is selected deterministically.
    - Otherwise, a weighted random selection is used among the local servers.
  - If no local mirrors are found, a weighted selection is performed among all valid servers.

- **Optional Rule Filtering:**
  - The current filtering of valid servers includes checks for availability, protocol support, and custom rules.
  - For debugging purposes, the rule filtering can be temporarily disabled to ensure all mirrors are taken into account.

## Impact

- **For Clients:**  
  Clients will now be more reliably directed to a local mirror when available, improving latency and download speed.

Please review the changes and let me know if further adjustments are needed.

SuperKali committed:
Improve mirror selection logic in Closest with full-server fallback and enhanced logging

SuperKali committed:
Remove debug log from code

SuperKali commented:
Hi @efectn,

I don't know if you've had a chance to look at my PR, let me know what you think.

Thank you.

efectn commented on servers.go:
```diff
@@ -189,105 +190,127 @@ type ComputedDistance struct {
-		choiceInterface = choices
-
-		r.serverCache.Add(scheme+"_"+ip.String(), choiceInterface)
+		r.serverCache.Add(cacheKey, choices)
```
I ain't sure about here. I think we should just put the selected one to the cache. Will []randutil.Choice be asserted to ComputedDistance?

efectn commented on servers.go:
```diff
@@ -189,105 +190,127 @@ type ComputedDistance struct {
+			Item:   item,
+		}
+	}
+	r.serverCache.Add(cacheKey, choices)
```
Here too

efectn commented on servers.go:
```diff
@@ -189,105 +190,127 @@ type ComputedDistance struct {
-		// Then, map them to distances from the client
-		c := lo.Map(validServers, func(server *Server, _ int) ComputedDistance {
-			distance := Distance(city.Location.Latitude, city.Location.Longitude, server.Latitude, server.Longitude)
+	const sameCityThreshold = 20000000.0 // 200 km
```
Maybe this can be configurable from yaml config.

SuperKali committed:
Add sameCityThreshold to config and always store ComputedDistance in cache

    - Introduce SameCityThreshold in the Config struct
    - Replace the local constant sameCityThreshold with the configurable value
    - Store a single ComputedDistance in the cache instead of []randutil.Choice
